### PR TITLE
EZP-30111: [REST] Added Content status to RestContent output

### PR DESCRIPTION
**Pull request by blankse**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30111](https://jira.ez.no/browse/EZP-30111)
| **Bug/Improvement**| no
| **New feature**    | yes
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

Currently I do not have the opportunity to know if the content is in the trash, right?
The code is similar to the VersionInfo output visitor

**TODO**:
- [x] Implement feature.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by andrerom** [comment]
👍
It will at least need a small change to REST spec in doc folder, but besides that maybe @ViniTou for instance knows :)

File: eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestContent.php [hunk 1]
```diff
@@ -190,4 +197,28 @@ public function visit(Visitor $visitor, Generator $generator, $data)
...
+
+            case ContentInfo::STATUS_PUBLISHED:
+                return 'PUBLISHED';
+
+            case ContentInfo::STATUS_TRASHED:
+                return 'TRASHED';
+        }
+
+        // @todo FIXME: What exception to use?
+        throw new \Exception('Undefined content status: ' . $status);
```
Comment (alongosz) [inline eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestContent.php]: Probably `BadStateException`

**Comment by blankse** [comment]
@alongosz @andrerom I rebased to master and changed the REST spec

**Comment by andrerom** [comment]
@alongosz you'd like to send this to QA?

**Comment by alongosz** [comment]
@andrerom yes

**Comment by blankse** [comment]
I got the following error in an PHP 5.6 environment (but in another PHP 5.6 environment it works):
```
Unexpected error, the message was : Compile Error: Cannot use eZ\Publish\Core\Base\Exceptions\BadStateException as BadStateException because the name is already in use in /vendor/datafactory/ezpublish-kernel/eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestContent.php on line 12
```

Should we use it with an alias?

`use eZ\Publish\Core\Base\Exceptions\BadStateException as BadStateExceptionException;`

Do you have a suggestion @andrerom @alongosz ?

**Comment by alongosz** [comment]
@blankse Kernel 7.x does not support PHP 5.6. You need 7.1 or higher.

**Comment by blankse** [comment]
@alongosz I have cherry picked this to a 6.13 branch in my fork.

If the problem at 7.1 definitely does not exist, we can leave it that way.

**Comment by alongosz** [comment]
@blankse maybe there's a cherry-pick issue there? I cherry-picked it right now to my local 6.13 and it looks rather correct. Maybe the fork you're cherry-picking onto has more changes than `ezsystems/ezpublish-kernel:6.13`? The error suggests that there are multiple imports of `BadStateException`. However this would be PHP version-independent, so that's a bit weird.

Does the error occur on syntax check:
```
php5.6 -l eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestContent.php
```
?

**Comment by blankse** [comment]
@alongosz No I use this fork also with PHP 7.2 and have no errors. On my local machine with PHP 5.6.39 it works also. But on one live system with PHP 5.6.34 generates this error.

The syntax check is everywhere:
`No syntax errors detected in eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestContent.php`

I think this PHP Version have a problem with the BadStateException Vistor in the same namespace:
https://github.com/ezsystems/ezpublish-kernel/tree/master/eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor

A alias like the VersionValue works with this. Do you think we should set one for BadStateException also to be on the safe side?
https://github.com/ezsystems/ezpublish-kernel/blob/55a062b7e7b479967101241506b6d4096f2d1c31/eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/RestContent.php#L14

**Comment by alongosz** [comment]
> I think this PHP Version have a problem with the BadStateException Vistor in the same namespace:
> https://github.com/ezsystems/ezpublish-kernel/tree/master/eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor
>

TBH this is weird, explicit `use` should hide it AFAIK. But maybe there's something I don't know about this.

> A alias like the VersionValue works with this. Do you think we should set one for BadStateException also to be on the safe side?

Ok, but then rather `as CoreBadStateException`.

**Comment by andrerom** [comment]
@alongosz Is there some way we can document all the possible values of this property in the spec?

**Comment by alongosz** [comment]
> @alongosz Is there some way we can document all the possible values of this property in the spec?

We need to ask @MagdalenaZuba. REST spec is being revamped for 3.0 AFAIR.

**Comment by MagdalenaZuba** [comment]
@alongosz  @andrerom  next step of our REST revamp is describing properties, so I believe we can

**Comment by andrerom** [comment]
@MagdalenaZuba So no blockers on your side for this then.

@alongosz In the meantime anyway we can describe this in the current REST spec for 2.5?

**Comment by alongosz** [comment]
Note: rebased because container crashed due to a lot of changes since last commit here ;)

**Comment by alongosz** [comment]
> @alongosz In the meantime anyway we can describe this in the current REST spec for 2.5?

@andrerom In the current REST doc this is specified in the [XSD schema for Common Definitions](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.0-rc2/doc/specifications/rest/REST-API-V2.rst#common-definitions) as the `versionStatus` type. While the name might be misleading, this is common for both the [Content XML Schema](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.0-rc2/doc/specifications/rest/REST-API-V2.rst#content-xml-schema) and the [Version XML Schema](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.0-rc2/doc/specifications/rest/REST-API-V2.rst#version-xml-schema).

There's no ideal place to add this info separately.

We could add it:
- in the table for [Creating content](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.0-rc2/doc/specifications/rest/REST-API-V2.rst#creating-content) because this is the first occurence,
- after [the examples for Creating Content](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.0-rc2/doc/specifications/rest/REST-API-V2.rst#id6),
- before [Managing Versions](https://github.com/ezsystems/ezpublish-kernel/blob/v7.5.0-rc2/doc/specifications/rest/REST-API-V2.rst#managing-versions) as a last section for Content-related doc.

**Comment by andrerom** [comment]
@alongosz No strong preferences on my side, what's yours?

**Comment by alongosz** [comment]
> @alongosz No strong preferences on my side, what's yours?

@andrerom probably at the end, before Managing Versions, but it doesn't make me 100% happy ;)

**Comment by micszo** [comment]
When item is in Trash and when it's removed permanently status is _TRASHED_. Is that alright?

**Comment by micszo** [comment]
Turns out this depends on server cache. After removing it the response is 404.

**Comment by alongosz** [comment]
@micszo when moved to trash, it should be _TRASHED_, when removed permanently it should be 404. If that's your case then everything is ok. Cache issues need to be solved separately, because they're not related to these changes.

**Comment by micszo** [comment]
Created "follow-up" ticket https://jira.ez.no/browse/EZP-30314.

**Pull request merged**